### PR TITLE
Improve order line discount discoverability and modal

### PR DESCRIPTION
**Pull request by mirekm**
image

image

https://github.com/user-attachments/assets/9a556627-b3f1-4f22-bff9-de736757eb48

### CODE REVIEW

File: src/orders/components/OrderDiscountModal/useDiscountForm.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+const numbersRegex = /^[0-9]*\.?[0-9]+$/;
+
+interface UseDiscountFormProps {
+  maxPrice: MoneyFragment;
+  existingDiscount?: OrderDiscountCommonInput;
+  isOpen?: boolean;
+}
+
+export const useDiscountForm = ({ maxPrice, existingDiscount, isOpen }: UseDiscountFormProps) => {
```
Comment (lkostrowski) [inline src/orders/components/OrderDiscountModal/useDiscountForm.ts]: can you re-write it to use react-hook-form? It grants you proper abstraction on forms, validations, submitting. No need to hold custom state etc.

You can keep the useDiscountForm hook and add `useForm` here for form management part

File: src/orders/components/OrderDiscountModal/OrderLineDiscountModal.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  lineData?: OrderLineData;
+  existingDiscount?: OrderDiscountCommonInput;
+  confirmStatus: ConfirmButtonTransitionState;
+  removeStatus: ConfirmButtonTransitionState;
+  onConfirm: (discount: OrderDiscountCommonInput) => void;
+  onRemove: () => void;
+  onClose: () => void;
+}
+
+export const OrderLineDiscountModal = ({
```
Comment (lkostrowski) [inline src/orders/components/OrderDiscountModal/OrderLineDiscountModal.tsx:49]: Let's add storybook stories for new components, that will help us with testing in isolation, testing PRs, visual regressions

You can easily prompt it "Add stories for `src/orders/components/OrderDiscountModal/OrderLineDiscountModal.tsx`". You can additionally describe specific cases you want to show (to diversify props). You can also prompt to write integration tests in storybook (eg. it can write you scenario that you can edit, save, confirm in sequence correctly)

File: src/orders/components/OrderLineMetadataDialog/OrderLineMetadataDialog.test.tsx [hunk 3]
```diff
@@ -107,9 +107,9 @@ describe("OrderLineMetadataDialog", () => {
     );
 
     // Assert
-    expect(screen.queryByText(mockData.quantity)).toBeInTheDocument();
-    expect(screen.queryByText(mockData.variant.name)).toBeInTheDocument();
-    expect(screen.queryByText(mockData.productSku)).toBeInTheDocument();
+    expect(screen.getByText(new RegExp(mockData.quantity.toString()))).toBeInTheDocument();
+    expect(screen.getByText(new RegExp(mockData.variant.name))).toBeInTheDocument();
+    expect(screen.getByText(new RegExp(mockData.productSku))).toBeInTheDocument();
```
Comment (lkostrowski) [inline src/orders/components/OrderLineMetadataDialog/OrderLineMetadataDialog.test.tsx]: Agree, try to assert by actual values, not regexp, it can lead to false positives

File: src/orders/components/OrderDiscountModal/messages.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    id: "IN5iJz",
+    defaultMessage: "Invalid value",
+    description: "value input helper text",
+  },
+  valueBiggerThatPrice: {
+    defaultMessage: "Cannot be higher than the price",
+    id: "VIdXPy",
+    description: "value input helper text",
+  },
+  valueBiggerThat100: {
```
Comment (lkostrowski) [inline src/orders/components/OrderDiscountModal/messages.ts]: that → than

File: src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx [hunk 5]
```diff
@@ -21,60 +12,25 @@ interface MoneyDiscountedCellProps {
   readonly locale: Locale;
 }
 
-const DATAGRID_BORDER_WIDTH = 1;
-const ROW_HEIGHT = cellHeight + DATAGRID_BORDER_WIDTH;
-
 export type MoneyDiscuntedCell = CustomCell<MoneyDiscountedCellProps>;
 
-const MoneyDiscountedCellEditor = ({ onFinishedEditing, value }) => {
```
Comment (lkostrowski) [inline src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx:29]: if you removed editor, how do we edit price now?

Comment (mirekm) [reply src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx:29]: In one unified modal.

File: src/orders/components/OrderDiscountModal/useDiscountForm.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    const isPercentage = calculationMode === DiscountValueTypeEnum.PERCENTAGE;
+    const topAmount = isPercentage ? 100 : maxAmount;
+    const parsedValue = parseNumericValue(value);
+
+    if (!numbersRegex.test(value)) {
+      return intl.formatMessage(messages.invalidValue);
+    }
+
+    if (parsedValue > topAmount) {
```
[the remark this answers is not in the document]
Comment (mirekm) [reply src/orders/components/OrderDiscountModal/useDiscountForm.ts:129]: This will be resolved once we merge https://github.com/saleor/saleor-dashboard/pull/6307

**Review by szczecha — DISMISSED** [review]

File: src/orders/components/OrderSummary/getUndiscountedSubtotal.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { type OrderLineFragment } from "@dashboard/graphql";
+
+type LineInput = Pick<OrderLineFragment, "undiscountedUnitPrice" | "quantity">;
+
+export function getUndiscountedSubtotal(lines: LineInput[], displayGrossPrices: boolean): number {
```
Comment (lkostrowski) [inline src/orders/components/OrderSummary/getUndiscountedSubtotal.ts]: add test

File: src/orders/components/OrderSummary/getUndiscountedSubtotal.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  return lines.reduce((sum, line) => {
+    const price = displayGrossPrices
+      ? line.undiscountedUnitPrice.gross.amount
+      : line.undiscountedUnitPrice.net.amount;
+
+    return sum + price * line.quantity;
```
Comment (lkostrowski) [inline src/orders/components/OrderSummary/getUndiscountedSubtotal.ts]: are you sure you want to calcualte this on the frontend? Saleor contains price calculation logic which is full of edge cases and actualy price may differ? Not to mention precision issues, normally there are libs for money calculations

File: src/orders/components/OrderSummary/getLineDiscountsSummary.test.ts [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+const makeLine = ({
+  unitDiscountAmount = 0,
+  quantity = 1,
+  discountType,
+}: {
+  unitDiscountAmount?: number;
+  quantity?: number;
+  discountType?: OrderDiscountType;
+}) => ({
```
Comment (lkostrowski) [inline src/orders/components/OrderSummary/getLineDiscountsSummary.test.ts]: put function return type (graphql fragment) here, so we have test in sync with expected data)

File: src/orders/components/OrderDiscountModal/OrderLineDiscountModal.tsx [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+  promotion: discountMessages.promotionSource,
+  other: discountMessages.genericSource,
+};
+
+const getDiscountSourceLabel = (
+  type: AutomaticDiscountInfo["type"],
+  intl: ReturnType<typeof useIntl>,
+): string => intl.formatMessage(discountSourceMessages[getDiscountTypeCategory(type)]);
+
+const formatDiscountSource = (
```
Comment (lkostrowski) [inline src/orders/components/OrderDiscountModal/OrderLineDiscountModal.tsx]: should be tested

File: src/orders/components/OrderDiscountModal/useDiscountForm.test.ts [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+      // Arrange
+      const { result } = renderHook(() => useDiscountForm({ maxPrice: defaultMaxPrice }));
+
+      // Act
+      act(() => {
+        result.current.setValue("value", "150");
+      });
+
+      // Assert
+      expect(result.current.valueErrorMsg).toBeTruthy();
```
Comment (lkostrowski) [inline src/orders/components/OrderDiscountModal/useDiscountForm.test.ts]: assert actual message, you may have other error than you think

File: src/products/components/OrderDiscountProviders/utils.ts [hunk 12]
```diff
@@ -55,3 +47,41 @@ export const getParsedDiscountData = ({
   calculationMode,
   reason,
 }: OrderDiscountCommonInput) => ({ reason, value, valueType: calculationMode });
+
+export const getAutomaticLineDiscounts = (
```
Comment (lkostrowski) [inline src/products/components/OrderDiscountProviders/utils.ts:51]: missing tests

File: src/products/components/OrderDiscountProviders/utils.ts [hunk 13]
```diff
@@ -19,34 +14,31 @@ export const useDiscountDialog = () => {
...
+  const line = order.lines.find(({ id }) => id === orderLineId);
+
+  if (!line) {
+    return null;
+  }
+
+  const manualDiscount = line.discounts?.find(d => d.type === OrderDiscountType.MANUAL);
+
+  if (!manualDiscount) {
     return null;
```
Comment (lkostrowski) [inline src/products/components/OrderDiscountProviders/utils.ts:33]: missing tests

File: src/products/components/OrderDiscountProviders/OrderDiscountProvider.tsx [hunk 14]
```diff
@@ -47,7 +50,13 @@ export const OrderDiscountProvider = ({ children, order }: OrderDiscountProvider
   const [orderDiscountRemove, orderDiscountRemoveOpts] = useOrderDiscountDeleteMutation({
     onCompleted: ({ orderDiscountDelete: { errors } }) => handleDiscountDataSubmission(errors),
   });
-  const handleDiscountDataSubmission = (errors: any[]) => {
+  const handleDiscountDataSubmission = async (errors: any[]) => {
```
Comment (lkostrowski) [inline src/products/components/OrderDiscountProviders/OrderDiscountProvider.tsx]: you should use `unknown` instead of `any`

File: src/orders/components/OrderSummary/getUndiscountedSubtotal.ts (hunk 8 above)
Comment (mirekm) [reply src/orders/components/OrderSummary/getUndiscountedSubtotal.ts]: I will simplify it to use `undiscountedTotalPrice` for now but to be discussed if we can quickly add `undiscountedSubtotal` to Order and Checkout.

File: .changeset/spotty-bikes-brush.md [hunk 15]
```diff
@@ -0,0 +1,5 @@
+---
+"saleor-dashboard": patch
+---
+
+Order-level and order-line discount flows now share the same UX in Draft and Unconfirmed orders. Access Add/Edit order line discount now also from the order row menu.
```
Comment (witoszekdev) [inline .changeset/spotty-bikes-brush.md:5]: suggestion: we could add screenshots and recording from this PR to changeset it would enchance our changelog nicely

File: src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx [hunk 16]
```diff
@@ -21,60 +12,25 @@ interface MoneyDiscountedCellProps {
...
-// TODO: add new design
 export const moneyDiscountedCellRenderer = (): CustomRenderer<MoneyDiscuntedCell> => ({
   kind: GridCellKind.Custom,
   isMatch: (c): c is MoneyDiscuntedCell => (c.data as any).kind === "money-discounted-cell",
   draw: (args, cell) => {
     const { ctx, theme, rect } = args;
     const { currency, value, undiscounted, locale } = cell.data;
     const hasValue = value === 0 ? true : !!value;
-    const formattedValue = getFormattedMoney(value, currency, locale, "-");
+    const formattedValue = getFormattedMoney(value ?? "", currency, locale, "-");
```
Comment (witoszekdev) [inline src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx]: question: if we don't know `value`, maybe we should just present empty string or `-`?


```suggestion
    const formattedValue = value ? getFormattedMoney(value, currency, locale, "-") : "";
```

File: src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx [hunk 17]
```diff
@@ -85,11 +41,4 @@ export const moneyDiscountedCellRenderer = (): CustomRenderer<MoneyDiscuntedCell
...
     return true;
   },
-  provideEditor: () => ({
-    editor: MoneyDiscountedCellEditor,
-    styleOverride: {
-      padding: `${ROW_HEIGHT}px 0 0 0`,
-    },
-    disableStyling: true,
-  }),
 });
```
Comment (witoszekdev) [inline src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx:47]: question: why do we remove this?

File: src/orders/components/OrderDiscountModal/DiscountFormFields.tsx [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+      },
+    ],
+    [intl],
+  );
+
+  return (
+    <Box display="flex" flexDirection="column" gap={3} overflow="hidden">
+      <RadioGroupField
+        choices={discountTypeChoices}
+        name="discountType"
```
Comment (witoszekdev) [inline src/orders/components/OrderDiscountModal/DiscountFormFields.tsx]: question: shouldn't this use `calculationMode` instead? since that's what this form field is actually controlling



nitpick: this form uses react-hook-form, so we also should have some existing wrapper for macaw components with react-hook-form handlers attached to them
```suggestion
        name="calculationMode"
```

one more thing: if we use hook form properly with `control` prop or with TypeScript generics we wouldn't need `as DiscountValueTypeEnum` assertion which breaks type safety

File: src/orders/components/OrderDiscountModal/DiscountFormFields.tsx [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+                className={styles.priceInput}
+                size="small"
+                type="number"
+                label={intl.formatMessage(messages.discountValueLabel)}
+                error={!!valueErrorMsg}
+                helperText={valueErrorMsg || ""}
+                value={toFixed(field.value, 2)}
+                onChange={field.onChange}
+                min="0"
+                step="0.01"
```
Comment (witoszekdev) [inline src/orders/components/OrderDiscountModal/DiscountFormFields.tsx]: issue: this might not be a correct step for each currency (not every one uses 2 decimal digits 😄) - afaik this is already handled in other dashboard components, so we should probably re-use logic from there

File: src/orders/components/OrderDiscountModal/DiscountModalBase.tsx [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+          overflow="hidden"
+        >
+          {header}
+          <Modal.Close>
+            <Button
+              data-test-id="close-button"
+              icon={<X size={20} />}
+              size="small"
+              variant="tertiary"
+              flexShrink="0"
```
[the remark this answers is not in the document]
Comment (witoszekdev) [reply src/orders/components/OrderDiscountModal/DiscountModalBase.tsx:80]: nitpick: good suggestion https://github.com/saleor/saleor-dashboard/pull/6434/changes#r2978287684

File: src/orders/components/OrderDiscountModal/OrderDiscountModal.stories.tsx
Comment (witoszekdev) [inline src/orders/components/OrderDiscountModal/OrderDiscountModal.stories.tsx:1]: tip: we can actually add tests here for checking interactions: https://storybook.js.org/docs/writing-tests/interaction-testing (Łukasz already added some in existing stories)

File: src/orders/components/OrderDiscountModal/useDiscountForm.test.ts
Comment (witoszekdev) [inline src/orders/components/OrderDiscountModal/useDiscountForm.test.ts:1]: question: don't we need to mock `react-intl` here for `useIntl` hook?

File: src/orders/components/OrderDiscountModal/formatDiscountSource.test.ts [hunk 21]
```diff
@@ -0,0 +1,96 @@
+import { OrderDiscountType } from "@dashboard/graphql";
+import { type AutomaticDiscountInfo } from "@dashboard/products/components/OrderDiscountProviders/types";
+import { type IntlShape, type MessageDescriptor } from "react-intl";
+
+import { formatDiscountSource } from "./formatDiscountSource";
+
+/** Jest maps `react-intl` to a mock whose formatMessage drops ICU values; use a minimal real formatter. */
+const createFormatMessageIntl = (): IntlShape =>
```
Comment (lkostrowski) [inline src/orders/components/OrderDiscountModal/formatDiscountSource.test.ts]: Can't you just use real intl for this test? If formatter heavily relies on intl patterns but you literally replace its implementation, you test nothing

**Review by lkostrowski — DISMISSED** [review]

File: .changeset/spotty-bikes-brush.md (hunk 15 above)
Comment (mirekm) [reply .changeset/spotty-bikes-brush.md:5]: Do you have an example up your sleeve so I could reference @witoszekdev ?

File: src/orders/components/OrderDiscountModal/DiscountFormFields.tsx (hunk 19 above)
Comment (mirekm) [reply src/orders/components/OrderDiscountModal/DiscountFormFields.tsx]: These steppers are not the right UX anyway - we have a PR to remove it from all the money field in fact. I will also remove it from here.

File: src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx (hunk 17 above)
Comment (mirekm) [reply src/components/Datagrid/customCells/Money/MoneyDiscountedCell.tsx:47]: I dropped the datagrid editor altogether and moved the discount editing to a regular modal - it's a better, more consistent primitive for our use case (the datagrid's editors make probably more sense in the "spreadsheet" scenario/context).

File: src/orders/components/OrderDiscountModal/DiscountFormFields.tsx (hunk 18 above)
Comment (mirekm) [reply src/orders/components/OrderDiscountModal/DiscountFormFields.tsx]: I built a new generic wrapper rather than inline `<Controller>` (in `components/HookFormRadioGroup/`).

File: src/orders/components/OrderDiscountModal/DiscountFormFields.tsx (hunk 19 above)
Comment (mirekm) [reply src/orders/components/OrderDiscountModal/DiscountFormFields.tsx]: Also no need for toFixed here - I dropped it.

File: src/orders/components/OrderDiscountModal/useDiscountForm.test.ts
Comment (mirekm) [reply src/orders/components/OrderDiscountModal/useDiscountForm.test.ts:1]: `react-intl` is already globally mocked for the whole suite via `moduleNameMapper` in jest.config.js

File: src/orders/components/OrderDiscountModal/OrderDiscountModal.stories.tsx
Comment (mirekm) [reply src/orders/components/OrderDiscountModal/OrderDiscountModal.stories.tsx:1]: Yep. Done.

**Review by lkostrowski — APPROVED** [review]
